### ISSUE TITLE
Crashing at different points for different files.

### ISSUE THREAD

**Issue by anderssnp** [rphlo/karttapullautin#133]
I have a couple of laz files I cant run. I tried both prebuilt and compiling the source, changing settings in the ini etc. No luck. here is an example output:

$ ./pullauta 24E003_639_71_7550.laz 
Karttapullautin v2.2.1
There is no warranty. Use it at your own risk!

Preparing input file
Done
Knoll detection part 1
Generating curves...
Killed

and here is another, slightly better, but still bad:

./pullauta 24E003_639_71_5050.laz 
Karttapullautin v2.2.1
There is no warranty. Use it at your own risk!

Preparing input file
Done
Knoll detection part 1
Generating curves...
Done
Knoll detection part 2
Detecting knolls...
Done
Contour generation part 1
Identifying knolls...
Done
Contour generation part 2
Generating curves...
Done
Contour generation part 3
Smooth curves...
Done
Contour generation part 4
Identifying dotknolls...
Done
Vegetation generation
Generating vegetation...
Done
Cliff generation
Identifying cliffs...
thread 'main' panicked at src/cliffs.rs:165:62:
called `Result::unwrap()` on an `Err` value: InvalidProbability
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Comment by antbern** [comment]
The last example there indicates that the value specified in the `cliffthin` configuration parameter is not a valid probability, i.e., the value is not in the range `0.0` to `1.0`.  Did you change any of the configuration parameters? 🤔  Also rerunning with the suggested `RUST_BACKRACE=1` will output more information which helps in debugging 🚀 

Also how big is the first file? If you get `Killed` it might be due to out-of-memory issues with your machine. If there is something in the code it will usually output informastion similar to your second example (eg. that the thread panicked).

**Comment by anderssnp** [comment]
Yes, setting `cliffthin` to 1.0 fixed the second problem, thanks! It was set to 1.15, but that came from the default generated ini file.

For the first file, it is still getting killed. The file is 47MB.

[pullauta.ini.txt](https://github.com/user-attachments/files/17028607/pullauta.ini.txt)

**Comment by antbern** [comment]
> Yes, setting `cliffthin` to 1.0 fixed the second problem, thanks! It was set to 1.15, but that came from the default generated ini file.

Great! The default in the currently generated default file is 1.0 (look at the `pullauta.default.ini` file here on GitHub) , so perhaps you had an old file laying around somehow 🤔 


> For the first file, it is still getting killed. The file is 47MB.

That is weird, such a small file is unlikely to cause out of memory errors. Could also be some kind of undefined behavior (eg. dereferencing a null pointer etc.) which causes the OS to kill the process. To figure that out we need to run through `valgrind` or similar sanitizer perhaps.

**Comment by antbern** [comment]
> For cliff thin parameter we could add safety check at parsing. And same for all probability values... @antbern 

That's  great idea @rphlo , with some nice error messages as well 💯

**Comment by anderssnp** [comment]
https://drive.google.com/file/d/1ZAJF_miC-_7DH5E2yccq3R1Cn-yeWlvs/view?usp=sharing
https://drive.google.com/file/d/1Ez89a8cOvJJJ1kulMFB0lRCYJWTpD049/view?usp=sharing

**Comment by antbern** [comment]
Just to confirm I checked the file on my computer, and it seems since the file is somehow corrupt, the code tries to allocate around 10GB of memory to hold some values for a computation. If that much memory is not available the program will most likely get killed (happens on my machine), which explains your sudden `Killed` output @anderssnp . 

Perhaps we should introduce some safety checks against excessive memory usage :thinking: @rphlo

**Comment by antbern** [comment]
I agree that limiting memory could be a bit subjective. There is also no real way to check how much memory will be allocated, so the only thing we can do is an estimation (eg. using the number of elements `N` * `size_of::<T>()` for a `Vec<T>` / `vec![T; N]` ) but we would have to do it in all places and the only reasonable thing would be to have this user controllable in the configuration with default to something that is generally reasonable (eg. the 20GB here should most likely fall outside such a default). But it might be too much work for not that much benefit. Issuing warnings that the user can see (with an estimation of the size) might be better. Then the user can choose to exit if they want...
